### PR TITLE
action_command_notification plugin - added add api

### PR DESCRIPTION
**Pull request by synman**
* [X ] Your changes are not possible to do through a plugin and relevant
    to a large audience (ideally all users of OctoPrint)
  * [N/A ] If your changes are large or otherwise disruptive: You have
    made sure your changes don't interfere with current development by
    talking it through with the maintainers, e.g. through a
    Brainstorming ticket
  * [maint ] Your PR targets OctoPrint's `devel` branch if it's a completely
    new feature, or `maintenance` if it's a bug fix or improvement of
    existing functionality for the current stable version (no PRs
    against `master` or anything else please)
  * [X ] Your PR was opened from a custom branch on your repository
    (no PRs from your version of `master`, `maintenance`, or `devel`
    please), e.g. `dev/my_new_feature` or `fix/my_bugfix`
  * [X ] Your PR only contains relevant changes: no unrelated files,
    no dead code, ideally only one commit - rebase and squash your PR
    if necessary!
  * [X ] Your changes follow the existing coding style
  * [N/A ] If your changes include style sheets: You have modified the
    `.less` source files, not the `.css` files (those are generated
    with `lessc`)
  * [X ] You have tested your changes (please state how!) - ideally you
    have added unit tests - created a client and called api with X-API-Key
  * [X ] You have run the existing unit tests against your changes and
    nothing broke
  * [N/A ] You have added yourself to the `AUTHORS.md` file :)

#### What does this PR do and why is it necessary?
Extend action_command_notification plugin to support an "add notification" api.  This will enable plugin developers an easy way to inject custom notifications into the main notifications sidebar

#### How was it tested? How can it be tested by the reviewer?
```
headers = {"contentType": "application/json; charset=UTF-8", "X-Api-Key": _plugin._settings.global_get(["api", "key"])}
data = {"command": "add", "message": "Machine Connected"}
response = requests.post("http://localhost/api/plugin/action_command_notification", headers=headers, json=data)

```

### CODE REVIEW

**Comment by synman** [comment]
apologies for the noise there.  I got things cleaned up and promise to run the pre commit checks moving forward

**Review by cp2004 — CHANGES_REQUESTED** [review]
Is there an example use case for this?

Plugins do not usually call OctoPrint's HTTP API, and it seems counter-intuitive to setup an API for plugins to use like this. The global API key is deprecated and will be removed (definitely by 2.0.0, but who knows when we will get there), so your code example would not work from within a plugin. It also requires plugins to track the port OctoPrint is running on.

Have you seen the [plugin helpers](https://docs.octoprint.org/en/master/plugins/helpers.html)? This would be a more sensible option to use, as it can be directly called from the Python side of a plugin, no need to do a HTTP round trip.

**Comment by foosel** [comment]
I'd also like to add that this particular plugin is meant to display *messages from the printer's firmware*. So opening that up for plugins to utilize screams misuse and potential abuse leading to (understandable) user confusion. I'd rather not see the scope of this plugin watered down and extended beyond its design goal. 

I feel this is a solution to a problem that has not been fully explained yet and that might have a way better solution.

**Comment by synman** [comment]
My thoughts are this valuable resource sitting on the sidebar is grossly underutilized.  Very few machines even bother to support it (https://community.octoprint.org/t/octoprint-tells-me-my-firmware-lacks-support-for-host-action-commands-what-does-this-mean/34588).  Case in point, //action: notification isn't even documented on the RepRap G-Code reference page.

I am not targeting action commands -- only notifications.

So why not open it up a little?  I have considered alternatives such as client side viewmodel manipulation or rewriting various status notifications via gcode hooks but this requires either  an active client session or loss of info being echoed in the terminal window.   

I believe the use case is still very viable but we are locked into a mechanism that is not widely known, or adopted.  6-7 years ago this was common knowledge for many of us and that is mainly why I still try to use it.

Here's an excerpt from a machine I'm targeting, for example:
```
Recv: [MSG:INFO: Compiled with ESP32 SDK:v4.4.1-1-gb8050b365e]
Recv: [MSG:INFO: Local filesystem type is spiffs]
Recv: [MSG:INFO: Configuration file:blackbox-x32.yaml]
Recv: [MSG:INFO: Machine BlackBox_X32_XYYZ-DOOR]
Recv: [MSG:INFO: Board Open Builds BlackBox X32]
Recv: [MSG:INFO: SPI SCK:gpio.18 MOSI:gpio.23 MISO:gpio.19]
Recv: [MSG:INFO: SD Card cs_pin:gpio.5 detect:NO_PIN]
Recv: [MSG:INFO: Stepping:RMT Pulse:2us Dsbl Delay:0us Dir Delay:1us Idle Delay:255ms]
Recv: [MSG:INFO: Axis count 3]
Recv: [MSG:INFO: Shared stepper disable gpio.17:low]
Recv: [MSG:INFO: Axis X (-505.000,0.000)]
Recv: [MSG:INFO:   Motor0]
Recv: [MSG:INFO:     standard_stepper Step:gpio.12 Dir:gpio.14 Disable:NO_PIN]
Recv: [MSG:INFO:  X Neg Limit gpio.35]
Recv: [MSG:INFO: Axis Y (-505.000,0.000)]
Recv: [MSG:INFO:   Motor0]
Recv: [MSG:INFO:     standard_stepper Step:gpio.27 Dir:gpio.26 Disable:NO_PIN]
Recv: [MSG:INFO:  Y Neg Limit gpio.34]
Recv: [MSG:INFO:   Motor1]
Recv: [MSG:INFO:     standard_stepper Step:gpio.33 Dir:gpio.32 Disable:NO_PIN]
Recv: [MSG:INFO: Axis Z (-105.000,0.000)]
Recv: [MSG:INFO:   Motor0]
Recv: [MSG:INFO:     standard_stepper Step:gpio.15 Dir:gpio.2:low Disable:NO_PIN]
Recv: [MSG:INFO:  Z Neg Limit gpio.39]
Recv: [MSG:INFO: safety_door_pin gpio.16:pu]
Recv: [MSG:INFO: Kinematic system: Cartesian]
Recv: [MSG:INFO: PWM Spindle Ena:gpio.13 Out:gpio.25 Dir:gpio.4 Freq:5000Hz Period:8191]
Recv: [MSG:INFO: Using spindle PWM]
Recv: [MSG:INFO: Mist coolant gpio.21]
Recv: [MSG:INFO: Probe Pin: gpio.22:low]
Recv: [MSG:INFO: Connecting to STA SSID:xxx]
Recv: [MSG:INFO: Connecting.]
Recv: [MSG:INFO: Connecting..]
Recv: [MSG:INFO: Connected - IP is xxxx]
Recv: [MSG:INFO: WiFi on]
Recv: [MSG:INFO: Start mDNS with hostname:http://blackbox-x32.local/]
Recv: [MSG:INFO: SSDP Started]
Recv: [MSG:INFO: HTTP started on port 80]
Recv: [MSG:INFO: Telnet started on port 23]
```
This is all very useful information that would normally be discarded.  The current method I use for handling this type of shunting to the notifications sidebar is rather ugly as it requires intercepting the message from the machine on gcode_received and putting it in a buffer array that I play back later by selectively dropping things like position / temperature updates / etc, all so I can mimic //action: notification

I imagine I'll just repackage the action_command_notification plugin inside my own otherwise, and disable the built-in one.  This is definitely one option but I was hoping to build upon something I've always thought was a cool built-in capability.

After all, even my 8 year old Wanhao Repetier machine flat out ignores it 99.5% of the time:

Screenshot 2022-12-20 at 4 04 23 PM

**Comment by synman** [comment]
Looking over the helpers documentation, I would be more than happy to retract this PR and resubmit with something like that instead.  This is obviously contingent with my intent not circumventing the roadmap.

File: src/octoprint/plugins/action_command_notification/__init__.py [hunk 1]
```diff
@@ -75,13 +82,19 @@ def on_api_get(self, request):
         )
 
     def get_api_commands(self):
-        return {"clear": []}
+        return {"clear": [], "add": []}
```
[the remark this answers is not in the document]
Comment (synman) [reply src/octoprint/plugins/action_command_notification/__init__.py:85]: Looking over the helpers documentation, I would be more than happy to retract this PR and resubmit with something like that instead. This is obviously contingent with my intent not circumventing the roadmap.

**Comment by synman** [comment]
I am going to kill this PR but do need to know if it's okay to proceed with an alternate method of using helpers.

**Pull request closed without merging**